src/main/java/rx/internal/operators/OperatorMerge.java

```diff
@@ -71,715 +72,878 @@
+                // mode has changed concurrently
+                // recheck state
+                m = mode;
+                // if in unbounded mode, quit
+                if (m == 0) {
+                    return;
+                } else
+                // if bounded mode, update request count and emit
+                if (m == 1) {
+                    BackpressureUtils.getAndAddRequest(this, n);
+                    subscriber.emit();
+                    return;
+                }
+                throw new IllegalStateException("Shouldn't be in mode: " + m);
             }
         }
-
-        private void handleNewSource(Observable<? extends T> t) {
-            if (childrenSubscribers == null) {
-                // lazily create this only if we receive Observables we need to subscribe to
-                childrenSubscribers = new SubscriptionIndexedRingBuffer<InnerSubscriber<T>>();
-                add(childrenSubscribers);
-            }
-            MergeProducer<T> producerIfNeeded = null;
-            // if we have received a request then we need to respect it, otherwise we fast-path
-            if (mergeProducer.requested != Long.MAX_VALUE) {
-                /**
-                 * <pre> {@code
-                 * With this optimization:
-                 * 
-                 * r.o.OperatorMergePerf.merge1SyncStreamOfN      1000  thrpt         5    57100.080     4686.331    ops/s
-                 * r.o.OperatorMergePerf.merge1SyncStreamOfN   1000000  thrpt         5       60.875        1.622    ops/s
-                 *  
-                 * Without this optimization:
-                 * 
-                 * r.o.OperatorMergePerf.merge1SyncStreamOfN      1000  thrpt         5    29863.945     1858.002    ops/s
-                 * r.o.OperatorMergePerf.merge1SyncStreamOfN   1000000  thrpt         5       30.516        1.087    ops/s
-                 * } </pre>
-                 */
-                producerIfNeeded = mergeProducer;
-            }
-            InnerSubscriber<T> i = new InnerSubscriber<T>(this, producerIfNeeded);
-            i.sindex = childrenSubscribers.add(i);
-            t.unsafeSubscribe(i);
-            if (!isUnsubscribed()) {
-                request(1);
-            }
+        public long produced(int n) {
+            return addAndGet(-n);
         }
+    }
+    
+    /**
+     * The subscriber that observes Observables. 
+     * @param <T> the value type
+     */
+    static final class MergeSubscriber<T> extends Subscriber<Observable<? extends T>> {
+        final Subscriber<? super T> child;
+        final boolean delayErrors;
+        final int maxConcurrent;
+        
+        MergeProducer<T> producer;
+        
+        /** Upstream might not produce scalar values at all so avoid creating a queue for it upfront. */
+        volatile RxRingBuffer scalarQueue;
+        /** Tracks the active subscriptions to sources. */
+        volatile CompositeSubscription subscriptions;
```

Comment:
The use of `CompositeSubscription` was specifically avoided before and replaced with `SubscriptionIndexedRingBuffer` to avoid the overhead of `HashSet` when constant add/remove is happening, which is a common behavior for `merge`.

Why did you not continue to use `SubscriptionIndexedRingBuffer`?